### DISCUSSION TITLE
Running a Python Script in MOOSE (via MultiApp?)

### DISCUSSION THREAD

**Discussion by wrkendrick** [idaholab/moose#16497] [Q&A General] [answered]
Hi all,

I'm currently running MOOSE as a heat transfer solver and I'm interested in adding in an existing script that solves a heat pipe environment. The issue is that the heat pipe script was originally FORTRAN and was recently, painstakingly, written into Python, so I have no interest in doing the same for C++.

I've seen some evidence of people utilizing Python-based scripts in MOOSE, but I can't find any clear examples or actual code of how you go about tying them together. I understand the basics of embedding Python into C++, but I'm not sure that's the right method, or if there's just some easy implementation via the MultiApp functionality.

I wanted to ask y'all for any advice or experiences you may have had on this sort of topic. Any help is greatly appreciated!

**Comment by aeslaughter** [comment, marked as answer]
You should probably create a [MOOSE Wrapped App](https://mooseframework.inl.gov/application_development/moose_wrapped_apps.html) that is capable of running the python via C++. 

We do have a heat pipe application at INL, but I am not sure it is being released outside of the lab. @joshuahansel can you comment on that?

**Reply by joshuahansel** [reply]
Sockeye (the MOOSE-based heat pipe application) does have users external to INL, but it's not open-source or anything - you'd need to apply for access. I'm not sure on what the criteria are (but probably include organization, project, and nationality). Let me you know if you'd like to try or if you'd like to learn more about Sockeye.

**Reply by permcody** [reply]
Sockeye is licensable: [You can apply here: www.inl.gov/ncrc](www.inl.gov/ncrc) 
Calling Python from C++ is possible, but you'll likely find yourself spending a lot of time working out the technical details. This gets further complicated when you consider that MOOSE is parallel. How exactly do you plan to perform parallel calculations in your Python code or do you just want to limit it to a single processor?

**Reply by wrkendrick** [reply]
> You should probably create a [MOOSE Wrapped App](https://mooseframework.inl.gov/application_development/moose_wrapped_apps.html) that is capable of running the python via C++.
> 
> We do have a heat pipe application at INL, but I am not sure it is being released outside of the lab. @joshuahansel can you comment on that?

Is there any more info / examples of using a Wrapped App in the MOOSE framework? I'm looking but having a hard time telling what specifically I need to create/pass. The wiki may be a little bugged, too, because when I look to see what variables I can use with ExternalObject I can't really tell: https://mooseframework.inl.gov/source/problems/ExternalProblem.html. It seems like a wrapped app is what I need, though.

> Sockeye is licensable: You can apply here: www.inl.gov/ncrc
>Calling Python from C++ is possible, but you'll likely find yourself spending a lot of time working out the technical details. This gets further complicated when you consider that MOOSE is parallel. How exactly do you plan to perform parallel calculations in your Python code or do you just want to limit it to a single processor?

That's a good point I hadn't really thought about. I think for the short term I can live on a single processor, I can sacrifice some runtime if it means that I can actually run.

**Reply by GiudGiud** [reply]
Hi

I'm sure there are openly accessible examples of external apps wrapped using ExternalProblem. I don't know them but you could ask for access to:
- Cardinal at ANL, with NEK5000 wrapped this way
- Warthog at ORNL / ANL, with Proteus [link](https://publications.anl.gov/anlpubs/2019/10/156480.pdf)
for an example.

Guillaume
